File: clients/ios/VisualPose/PoseMapper.swift
import Foundation
import UIKit

protocol PoseLandmarkPoint {
    var x: Float { get }
    var y: Float { get }
    var z: Float { get }
    var visibility: Float? { get }
}

struct LandmarkXYZ: PoseLandmarkPoint {
    var x: Float
    var y: Float
    var z: Float
    var visibility: Float?
}

protocol PoseMarker: AnyObject {
    func detect(_ image: UIImage) -> [[any PoseLandmarkPoint]]
    func detectForVideo(_ image: UIImage, timestampMs: Int) -> [[any PoseLandmarkPoint]]
}

struct MappedPeople {
    var poses: [CocoPose]
    var blazePeople: [[BlazeJoint]]
    var bboxes: [BBox]
}

enum PoseMapper {
    static func fromLandmarks(
        _ poses: [[any PoseLandmarkPoint]],
        width: Int,
        height: Int,
        device: String
    ) -> MappedPeople {
        let w = Float(max(width, 1))
        let h = Float(max(height, 1))
        let n = poses.count
        var flat = [Float](repeating: 0, count: n * 33 * 4)
        var blazePeople: [[BlazeJoint]] = []
        var o = 0
        for person in poses {
            var joints: [BlazeJoint] = []
            joints.reserveCapacity(33)
            for i in 0..<33 {
                if i >= person.count {
                    joints.append(BlazeJoint(x: 0, y: 0, z: 0, confidence: 0))
                    o += 4
                    continue
                }
                let lm = person[i]
                let x = lm.x * w
                let y = lm.y * h
                let z = lm.z * w
                let vis = lm.visibility ?? 1
                joints.append(BlazeJoint(x: x, y: y, z: z, confidence: vis))
                flat[o] = x; o += 1
                flat[o] = y; o += 1
                flat[o] = z; o += 1
                flat[o] = vis; o += 1
            }
            blazePeople.append(joints)
        }
        let json = CoreMap.fromBlaze33(
            xyzVis: flat,
            numPeople: n,
            width: max(width, 1),
            height: max(height, 1),
            latencyMs: 0,
            device: device
        )
        return MappedPeople(
            poses: CoreJson.poses(json),
            blazePeople: blazePeople,
            bboxes: CoreJson.bboxes(json)
        )
    }
}

File: clients/android/app/src/main/java/local/visual/corepose/PoseMapper.kt
package local.visual.corepose

import com.google.mediapipe.tasks.components.containers.NormalizedLandmark

data class MappedPeople(
    val poses: List<CocoPose>,
    val blazePeople: List<List<BlazeJoint>>,
    val bboxes: List<BBox>,
)

object PoseMapper {
    fun fromLandmarks(
        poses: List<List<NormalizedLandmark>>,
        width: Int,
        height: Int,
        device: String,
    ): MappedPeople {
        val w = width.coerceAtLeast(1).toFloat()
        val h = height.coerceAtLeast(1).toFloat()
        val n = poses.size
        val flat = FloatArray(n * 33 * 4)
        val blazePeople = ArrayList<List<BlazeJoint>>(n)
        var o = 0
        for (person in poses) {
            val joints = ArrayList<BlazeJoint>(33)
            for (i in 0 until 33) {
                val lm = person.getOrNull(i)
                if (lm == null) {
                    joints.add(BlazeJoint(0f, 0f, 0f, 0f))
                    o += 4
                    continue
                }
                val x = lm.x() * w
                val y = lm.y() * h
                val z = lm.z() * w
                val vis = lm.visibility().orElse(1f)
                joints.add(BlazeJoint(x, y, z, vis))
                flat[o++] = x
                flat[o++] = y
                flat[o++] = z
                flat[o++] = vis
            }
            blazePeople.add(joints)
        }
        val json = CoreMap.fromBlaze33(flat, n, width.coerceAtLeast(1), height.coerceAtLeast(1), 0f, device)
        return MappedPeople(
            poses = CoreJson.poses(json),
            blazePeople = blazePeople,
            bboxes = CoreJson.bboxes(json),
        )
    }
}
